refactor(compiler): retrieve variables from context inside nested template listener

crisbeto opened the pull request:
This is a prerequisite for #40360. Given the following template which has a listener that references a variable from a parent template (`name`):

```
<ng-template let-name="name">
  <button (click)="hello(name)"></button>
</ng-template>
```

We generate code that looks that looks like the following. Note how we access `name` through `ctx`:

```js
function template(rf, ctx) {
  if (rf & 1) {
    const r0 = ɵɵgetCurrentView();
    ɵɵelementStart(0, "button", 2);
    ɵɵlistener("click", function() {
      ɵɵrestoreView(r0);
      const name_r0 = ctx.name; // Note the `ctx.name` access here.
      const ctx_r1 = ɵɵnextContext();
      return ctx_r1.log(name_r0);
    });
    ɵɵelementEnd();
  }
}
```

This works fine at the moment, because the template context object can't be changed after creation. The changes in #40360 allow for the object to be changed, which means that the `ctx` reference inside the listener will be out of date, because it was bound during creation mode.

This PR aims to address the issue by accessing the context inside listeners through the saved view reference. With the new code, the generated code from above will look as follows:

```js
function template(rf, ctx) {
  if (rf & 1) {
    const r0 = ɵɵgetCurrentView();
    ɵɵelementStart(0, "button", 2);
    ɵɵlistener("click", function() {
      const restoredCtx = ɵɵrestoreView(r0);
      const name_r0 = restoredCtx.name;
      const ctx_r1 = ɵɵnextContext();
      return ctx_r1.log(name_r0);
    });
    ɵɵelementEnd();
  }
}
```

crisbeto commented on packages/compiler-cli/test/compliance_old/r3_view_compiler_template_spec.ts:
```diff
@@ -56,7 +56,7 @@ describe('compiler compliance: template', () => {
           $i0$.ɵɵlistener("click", function MyComponent_ul_0_li_1_div_1_Template_div_click_0_listener(){
             $i0$.ɵɵrestoreView($s$);
-            const $inner$ = ctx.$implicit;
+            const $inner$ = $s$[8].$implicit;
```
Note for reviewer: we might actually want to have a new instruction like `getContext(r0)` since it would allow for better stack traces and would be more consistent. I used the direct array access for now since it's a bit smaller and I wanted to verify that this is the approach we want to take.

JoostK replied:
I think something like `getContext` would indeed follow the instruction paradigm, in contrast to this direct array access into the runtime data structures. Since we're no longer planning on treating instructions as public API I don't think it makes much of a difference, although I have a slight preference towards having an instruction for this as well for consistency.

@mhevery should probably weigh in here as well.

mhevery replied:
Yes, please add new instruction. Even though it is not public API it still shows up in stack traces.

crisbeto commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -358,8 +358,14 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
-            rhs = o.variable(CONTEXT_NAME);
+            if (scope.isListenerScope()) {
+              // Either ctx or r0[CONTEXT] where r0 = getCurrentView();
+              rhs = scope.getRestoreViewVariable()?.key(o.literal(core.CONTEXT)) ||
+                  o.variable(CONTEXT_NAME);
```
I haven't added any new unit tests since this these changes don't affect the current runtime code. I'll add more in #40360, if this PR lands.

AndrewKushnir replied:
I believe we can use `??` here:
```suggestion
              rhs = scope.getRestoreViewVariable()?.key(o.literal(core.CONTEXT)) ??
                  o.variable(CONTEXT_NAME);
```

atscott commented on packages/compiler/src/core.ts:
```diff
@@ -550,3 +550,7 @@ export const enum AttributeMarker {
+
+// Pasted from render3/interfaces/view.ts
+/** Index at which the context of an LView is stored. */
+export const CONTEXT = 8;
```
nit: does this need to be exported?

crisbeto replied:
It doesn't need to be exported since it's only used by the compiler to generate a number literal.

atscott reviewed (approved):
LGTM, thanks Kristiyan! I think it would be good to get someone more familiar with this area of the code to review as well.

AndrewKushnir commented on packages/compiler/src/core.ts:
```diff
@@ -550,3 +550,7 @@ export const enum AttributeMarker {
   I18n = 6,
 }
+
+// Pasted from render3/interfaces/view.ts
+/** Index at which the context of an LView is stored. */
```
Just a small proposal on docs:
```suggestion
/**
 * Index at which the context of an LView is stored.
 * Note: it **must** be in sync with corresponding const in packages/core/src/render3/interfaces/view.ts.
 */
```

AndrewKushnir commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -358,8 +358,14 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
-            // e.g. ctx
-            rhs = o.variable(CONTEXT_NAME);
+            if (scope.isListenerScope()) {
+              // Either ctx or r0[CONTEXT] where r0 = getCurrentView();
```
I think it'd be great to extend the comment a bit to cover the differences between `ctx` and `r0[CONTEXT]` and why we generate one vs another in some cases. We can refer to #40360 as well for additional context (even though that PR has not landed yet).

AndrewKushnir reviewed (commented):
Thanks for the fix @crisbeto 👍 Just left a few minor comments.

JoostK reviewed (approved):
LGTM!

AndrewKushnir reviewed (approved):
@crisbeto thanks for updating the PR and introducing an instruction 👍
This looks very clean and we don't have to sync constant values between core/compiler.

Quick question: you've added "target: major" label, so I'm wondering if there is any risk that you see and this change should not go into patch branch?

crisbeto commented:
I think it's a fairly safe change, but I set it to `major`, because it's meant to go together with #40360 which is a bit more risky.

mhevery commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_template/nested_template_context_many_bindings_template.js:
```diff
@@ -4,8 +4,8 @@ function MyComponent_div_0_Template(rf, ctx) {
       $r3$.ɵɵrestoreView($s$);
-      const $d$ = ctx.$implicit;
-      const $i$ = ctx.index;
+      const $d$ = $r3$.ɵɵgetContext($s$).$implicit;
+      const $i$ = $r3$.ɵɵgetContext($s$).index;
```
```suggestion
      const ctx = $r3$.ɵɵgetContext($s$);
      const $d$ = ctx.$implicit;
      const $i$ = ctx.index;
```

We should store it in local var, rather than creating the call instruction multiple times.

mhevery replied:
Come to think of it can we just change `ɵɵrestoreView` to return `ctx`
```
  const ctx = $r3$.ɵɵrestoreView($s$);
  const $d$ = ctx.$implicit;
  const $i$ = ctx.index;
```

mhevery commented on packages/compiler-cli/test/compliance_old/r3_view_compiler_template_spec.ts:
```diff
@@ -151,8 +151,8 @@ describe('compiler compliance: template', () => {
               $r3$.ɵɵrestoreView($s$);
-              const $d$ = ctx.$implicit;
-              const $i$ = ctx.index;
+              const $d$ = $r3$.ɵɵgetContext($s$).$implicit;
+              const $i$ = $r3$.ɵɵgetContext($s$).index;
```
```suggestion
              const ctx = $r3$.ɵɵgetContext($s$);
              const $d$ = ctx.$implicit;
              const $i$ = ctx.index;
```

crisbeto commented:
Let's abstain from merging it in now until I can change it so it declares a variable for the context. It would still be good to know if the current presubmit revealed any issues.

crisbeto commented:
@mhevery I've reworked it to declare a variable instead of going through `getContext`.

crisbeto committed:
refactor(compiler): retrieve variables from context inside nested template listener

This is a pre-requisite for #40360. Given the following template which has a listener
that references a variable from a parent template (`name`):

```
<ng-template let-name="name">
  <button (click)="hello(name)"></button>
</ng-template>
```

We generate code that looks that looks like. Note how we access `name` through `ctx`:

```js
function template(rf, ctx) {
  if (rf & 1) {
    const r0 = ɵɵgetCurrentView();
    ɵɵelementStart(0, "button", 2);
    ɵɵlistener("click", function() {
      ɵɵrestoreView(r0);
      const name_r0 = ctx.name; // Note the `ctx.name` access here.
      const ctx_r1 = ɵɵnextContext();
      return ctx_r1.log(name_r0);
    });
    ɵɵelementEnd();
  }
}
```

This works fine at the moment, because the template context object can't be changed after creation.
The changes in #40360 allow for the object to be changed, which means that the `ctx` reference
inside the listener will be out of date, because it was bound during creation mode.

This PR aims to address the issue by accessing the context inside listeners through the saved
view reference. With the new code, the generated code from above will look as follows:

```js
function template(rf, ctx) {
  if (rf & 1) {
    const r0 = ɵɵgetCurrentView();
    ɵɵelementStart(0, "button", 2);
    ɵɵlistener("click", function() {
      const restoredCtx = ɵɵrestoreView(r0);
      const name_r0 = restoredCtx.name;
      const ctx_r1 = ɵɵnextContext();
      return ctx_r1.log(name_r0);
    });
    ɵɵelementEnd();
  }
}
```

mhevery commented:
[presubmit](http://test/OCL:357972865:BASE:357972898:1613584919668:f1d633bd)
